diff --git a/src/server/game/Entities/Player/Player.cpp b/src/server/game/Entities/Player/Player.cpp
index 630551c..fad4682 100755
--- a/src/server/game/Entities/Player/Player.cpp
+++ b/src/server/game/Entities/Player/Player.cpp
@@ -17,6 +17,7 @@
  */

 #include "Common.h"
+#include "Config.h"
 #include "Language.h"
 #include "DatabaseEnv.h"
 #include "Log.h"
@@ -7381,6 +7382,54 @@ void Player::UpdateZone(uint32 newZone, uint32 newArea)
         SendInitWorldStates(newZone, newArea);              // only if really enters to new zone, not just area change, works strange...
     }

+    // Prevent players from accessing GM Island
+    if (sWorld->getBoolConfig(CONFIG_GMISLAND_PLAYERS_NOACCESS_ENABLE) == true)
+    {
+        if (newZone == 876 && AccountMgr::IsPlayerAccount(GetSession()->GetSecurity()))
+        {
+            uint32 map = 13;
+            double coords[4] = { 1.118799, 0.477914, -144.708650, 3.133046 }; // Jail Box
+
+            std::string config = ConfigMgr::GetStringDefault("GMIsland.TeleportTo", "");
+            if (config != "")
+            {
+                std::istringstream ss(config);
+
+                for (int i = 0; i < 5; ++i)
+                {
+                    if (!ss)
+                        break;
+
+                    if (i == 0)
+                        ss >> map;
+                    else
+                        ss >> coords[i - 1];
+                }
+            }
+
+            if (map == 876)
+            {
+                sLog->outError("Error: Cannot set tele to GM Island (map: 876). Sending possible hacker to default location. (Jail Box)");
+                map = 13;
+                coords[0] = 1.118799;       // x
+                coords[1] = 0.477914;       // y
+                coords[2] = -144.708650;    // z
+                coords[3] = 3.133046;       // orientation
+            }
+
+            sLog->outError("Player (GUID: %u) tried to access GM Island.", GetGUIDLow());
+            TeleportTo(map, coords[0], coords[1], coords[2], coords[3]); // Tele to Jail Box
+            if (map == 13)
+                CastSpell(this, 9454, true); // Cast GM Freeze on player
+
+            if (sWorld->getBoolConfig(CONFIG_GMISLAND_BAN_ENABLE) == true)
+            {
+                sLog->outError("Player (GUID: %u) tried to access GM Island. Banning Player Account.", GetGUIDLow());
+                sWorld->BanAccount(BAN_ACCOUNT, this->GetName(), secsToTimeString(TimeStringToSecs("-1"), true).c_str(), "Being on GM Island", "Anticheat protection");
+            }
+        }
+    }
+
     // group update
     if (GetGroup())
     {
diff --git a/src/server/game/Spells/SpellEffects.cpp b/src/server/game/Spells/SpellEffects.cpp
index ef5f32b..8f7d8f2 100755
--- a/src/server/game/Spells/SpellEffects.cpp
+++ b/src/server/game/Spells/SpellEffects.cpp
@@ -5659,6 +5659,13 @@ void Spell::EffectStuck(SpellEffIndex /*effIndex*/)

     Player* target = (Player*)m_caster;

+    // Prevent players from trying to unstuck themselves in the Jail box.
+    if (target->GetMapId() == 13 && AccountMgr::IsPlayerAccount(target->GetSession()->GetSecurity()))
+    {
+        sLog->outError("Player %s (guid %u) tried to use unstuck in Jail box.", target->GetName(), target->GetGUIDLow());
+        return;
+    }
+
     sLog->outDebug(LOG_FILTER_SPELLS_AURAS, "Spell Effect: Stuck");
     sLog->outDetail("Player %s (guid %u) used auto-unstuck future at map %u (%f, %f, %f)", target->GetName(), target->GetGUIDLow(), m_caster->GetMapId(), m_caster->GetPositionX(), target->GetPositionY(), target->GetPositionZ());

diff --git a/src/server/game/World/World.cpp b/src/server/game/World/World.cpp
index 354b36b..4591f73 100755
--- a/src/server/game/World/World.cpp
+++ b/src/server/game/World/World.cpp
@@ -1191,6 +1191,10 @@ void World::LoadConfigSettings(bool reload)
     m_bool_configs[CONFIG_PDUMP_NO_PATHS] = ConfigMgr::GetBoolDefault("PlayerDump.DisallowPaths", true);
     m_bool_configs[CONFIG_PDUMP_NO_OVERWRITE] = ConfigMgr::GetBoolDefault("PlayerDump.DisallowOverwrite", true);

+    // Prevent players from accessing GM Island
+    m_bool_configs[CONFIG_GMISLAND_PLAYERS_NOACCESS_ENABLE] = ConfigMgr::GetBoolDefault("GMIsland.PlayersNoAccess.Enable", true);
+    m_bool_configs[CONFIG_GMISLAND_BAN_ENABLE] = ConfigMgr::GetBoolDefault("GMIsland.Ban.Enable", false);
+
     sScriptMgr->OnConfigLoad(reload);
 }

diff --git a/src/server/game/World/World.h b/src/server/game/World/World.h
index be4c412..be4692f 100755
--- a/src/server/game/World/World.h
+++ b/src/server/game/World/World.h
@@ -163,6 +163,8 @@ enum WorldBoolConfigs
     CONFIG_QUEST_IGNORE_AUTO_ACCEPT,
     CONFIG_QUEST_IGNORE_AUTO_COMPLETE,
     CONFIG_WARDEN_ENABLED,
+    CONFIG_GMISLAND_PLAYERS_NOACCESS_ENABLE,
+    CONFIG_GMISLAND_BAN_ENABLE,
     BOOL_CONFIG_VALUE_COUNT
 };

diff --git a/src/server/worldserver/worldserver.conf.dist b/src/server/worldserver/worldserver.conf.dist
index fae38bd..eb163a2 100644
--- a/src/server/worldserver/worldserver.conf.dist
+++ b/src/server/worldserver/worldserver.conf.dist
@@ -2801,4 +2801,27 @@ PlayerDump.DisallowPaths = 1
 PlayerDump.DisallowOverwrite = 1

 #
+#      GMIsland.PlayersNoAccess.Enable
+#        Prevent access for players to gm island
+#        Default: 1 - Enable
+#                 0 - Disable
+
+GMIsland.PlayersNoAccess.Enable = 1
+
+#
+#      GMIsland.TeleportTo
+#        Teleport player to this coordinates (map, x, y, z, orientation)
+#        Default: "13 1.118799 0.477914 -144.708650 3.133046" (Jailbox)
+
+GMIsland.TeleportTo = "13 1.118799 0.477914 -144.708650 3.133046"
+
+#
+#      GMIsland.Ban.Enable
+#        Enable to ban players that tries to access GM Island without proper Access
+#        Default: 0 - Disable
+#                 1 - Enable
+
+GMIsland.Ban.Enable = 0
+
+#
 ###################################################################################################
